pkg/relay/ethereum/tx.go

```diff
@@ -56,19 +59,24 @@ func (c *Chain) SendMsgs(msgs []sdk.Msg) ([]core.MsgID, error) {
 		// case *transfertypes.MsgTransfer:
 		// 	err = c.client.transfer(msg)
 		default:
+			logger.Error("failed to send msg", errors.New("illegal msg type"), "msg", msg)
 			panic("illegal msg type")
 		}
 		if err != nil {
+			logger.Error("failed to send msg", err, "msg", msg)
 			return nil, err
 		}
 		if receipt, revertReason, err := c.client.WaitForReceiptAndGet(ctx, tx.Hash(), c.config.EnableDebugTrace); err != nil {
+			logger.Error("failed to get receipt", err, "msg", msg)
 			return nil, err
 		} else if receipt.Status == gethtypes.ReceiptStatusFailed {
-			return nil, fmt.Errorf("tx execution failed: revertReason=%s, msgIndex=%d, msg=%v", revertReason, i, msg)
+			err := fmt.Errorf("tx execution failed: revertReason=%s, msgIndex=%d, msg=%v", revertReason, i, msg)
+			logger.Error("tx execution failed", err, "msg", msg)
+			return nil, err
```

Comment:
@dongrie Ah. Sorry, but my intention was/is the following...
```golang
err := fmt.Errorf("tx execution failed: revertReason=%s, msgIndex=%d, msg=%v", revertReason, i, msg)
logger.Error("tx execution failed", err, "revert_reason", revertReason, "msg_index", msgIndex, "msg", msg)
return nil, err
```

In my understanding, when structured logging is used, users of logs always obtain information from attributes  and never parse free-form error strings.
Therefore any important information (e.g. revertReason, msgIndex, msg) should be embedded as attributes.
My code above looks redundant but redundancy is better than lack of information.

Reply:
https://github.com/datachainlab/ethereum-ibc-relay-chain/pull/13/commits/bcfa6f2e5e0b6e20c47c3a46398c25faee2c7b43
Change to use structured logging